### ISSUE TITLE
Gradient Boosted Classification Malloc Errors

### ISSUE THREAD

**Issue by raoberman** [IntelPython/daal4py#123]
I found 3 errors when running gradient boosted classification (the same errors may also come up with gradient boosted regression, which acts in a similar behavior).  These bugs are inconsistent behavior but appear frequently when I run my code (with at least one error appearing about 60% of the time):

When I run my data (Pandas DataFrames) through the gradient boosted classifier training, 9/10 times my code will break and result in a malloc unsigned long error (see first image).

![malloclong](https://user-images.githubusercontent.com/26121125/61091460-1240d880-a3f7-11e9-9307-53c847d04360.PNG)

While trying to do more testing with this bug. I found a different inconsistent error while running the same exact code - this time saying there was a corrupted double-linked list somewhere in memory (see second image).

![linkedlist](https://user-images.githubusercontent.com/26121125/61091895-c98a1f00-a3f8-11e9-9f07-e92d0c6b676a.PNG)

When I tried to troubleshoot this issue by double-checking that my data was in a valid format and re-casting the data as pandas DataFrames, I found another inconsistent error saying that there was a malloc memory corruption error (see third image).

![malloccorruption](https://user-images.githubusercontent.com/26121125/61091730-16212a80-a3f8-11e9-813c-f26aebfec6c1.PNG)

 I have already checked that there is no issue with the data itself, its size, or the code I run before it (i.e. the data is valid and should be accepted) - the issue is within gradient boosted classification. 

Within the attached mallocproblems.zip zipfile: 

[mallocproblems.zip](https://github.com/IntelPython/daal4py/files/3387008/mallocproblems.zip)

•	**Daal4py_solution_11_place.py** – The code that causes the malloc error when I run it. This is a shortened version that isolates the issue to the line causing the issue, which is the gradient boosted classifier training. A successful run will print the time in seconds it takes for the code to initialize centroids and train the data for KMeans and put it through a gradient boosted classifier.
•	**D4p_cluster_target_encoder.py** – A helper class that encodes the training data, initializes centroids for the KMeans clustering model, then trains the KMeans model. This runs fine.
•	**Train.csv** –  The training data for this workflow coding example. (4209 rows with 378 columns)
•	**Test.csv** – The test data for this workflow coding example (not used in the shortened daal4py_solution_11_place.py, but used for prediction)
•	**Df_classification_train.csv** – the training data used for the original gradient boosted classification validation test
•	**Df_classification_test.csv** – the test data used for the original gradient boosted classification validation test
•	**Gradient_boosted_classification_batch.py** – the original validation test code. This works perfectly fine.
•	**Problem_gradient_boosted_classification.py** – the validation test, with the only modification being the inclusion of the Kaggle data (and the KMeans clustering encoder) instead of the validation test’s original datasets. This originally revealed the inconsistent linked list and malloc unsigned errors.
•	**Inconsistent_malloclong_and_list.py** – Very similar to problem_gradient_boosted_classification.py, but rather than the first parameter including np.number, it is including np.int_ (the 1s and 0s are casted as int). This code tends to make the linked list and malloc unsigned errors show up more frequently.
•	**Malloccorruption.py** – Based on problem_gradient_boosted.py, this version has the first parameter does [include =(np.int_)] and then casts the entire dataframe to float, and the second parameter is changed from a pandas Series to a pandas DataFrame casted to float. Thus, the labels and training features are the same exact type of object (pandas dataframes casted to float), but interestingly enough  this produces a malloc memory corruption error.

**Comment by fschlimb** [comment]
@raoberman Does this also occur with numpy arrays (e.g. when passing df.values)?

**Comment by raoberman** [comment]
Yes, see the attached file and screenshots (which converts the data to numpy arrays before being processed by the gb classification training).

[numpy_gb.zip](https://github.com/IntelPython/daal4py/files/3387106/numpy_gb.zip)

![numpymalloclong](https://user-images.githubusercontent.com/26121125/61140852-1d404b00-a481-11e9-9a1a-29456d5723d4.PNG)

![numpymalloclinkedlist](https://user-images.githubusercontent.com/26121125/61140864-2204ff00-a481-11e9-8597-26214241c9ef.PNG)

**Comment by masdevas** [comment]
Hello. This bug has been fixed in DAAL side and will be released in DAAL2019u5 release. The problem was in preprocessing function, which created too much bins for data items and overflowed bins storage.

**Comment by oleksandr-pavlyk** [comment]
Indeed, with daal 2019.5 bits, using daal4py obtained from `conda create -n test-fix -c intel daal4py=2019.5 pandas scipy`, I have

```
(test-fix) [10:51:04 vmlin mallocproblems]$ python -c "import daal4py; print((daal4py.__daal_run_version__, daal4py.__daal_link_version__))"
("20190005_b'505204e8'", '20190005_20190730')
(test-fix) [10:51:08 vmlin mallocproblems]$ python daal4py_solution_11_place.py
Time in seconds:  0.49596571922302246
```

**Comment by oleksandr-pavlyk** [comment]
Closing as fixed.
